### DISCUSSION TITLE
GSOC 2026 : Project #6 Develop an OpenVINO-Domain Specialized Coder Model with SFT/GRPO/RAG

### DISCUSSION THREAD

**Discussion by Shehrozkashif** [openvinotoolkit/openvino#34615] [Google Summer of Code]
## GSoC 2026: Project 6 – Develop an OpenVINO-Domain Specialized Coder Model with SFT/GRPO/RAG

Hello @7taozhou7, @yinquan251 and everyone! Hope you're doing well.

I'm **Shehroz Kashif**, a **Software Engineering graduate from UIT University**. Over the past several years, I have developed a strong foundation in **systems programming, computer architecture, and AI systems**, with hands-on experience spanning hardware and software.

My journey began with **C, C++, and Java**, and later expanded into **computer architecture and hardware design** while working at the **Micro Electronics Research Lab (MERL)**. There, I explored **RISC-V ISA**, designed a **5-stage pipelined RV32I processor ([Vermithor](https://github.com/Shehrozkashif/Vermithor))** using **Scala and Chisel**, and gained practical experience in low-level system design and performance optimization.

I also served as an **LFX mentee at RISC-V International**, where I contributed to the **[RISC-V Unified Database (UDB)](https://github.com/riscv/riscv-unified-db)** under mentorship from engineers at **Qualcomm, Ventana, and Synopsys**. My contributions included multiple **merged PRs** such as fixes for operand handling, compressed instructions, and YAML schema corrections. This experience gave me insight into **collaborating in large open-source ecosystems** and delivering high-quality contributions.

Over the past few years, my focus shifted to **machine learning and LLMs**, culminating in my **final year project [AI4Org](https://github.com/merledu/ai4org)**, a **GAN-inspired hallucination reduction framework for LLMs**. I have fine-tuned models like **TinyLlama, GPT-2, Llama-2, and Mistral-7B** on consumer GPUs, experimenting with **SFT, PEFT (LoRA), RAG, and reinforcement learning approaches**. Along the way, I explored **memory-efficient training techniques** such as quantization, gradient checkpointing, and mixed precision.

Recently, I have been contributing to the **OpenVINO ecosystem**, especially **NNCF**, and I am keen to expand my contributions while building a **domain-specialized coder model for OpenVINO** using **SFT, GRPO, and RAG**.

I have **drafted my proposal and emailed it to you**, and I would appreciate it if you could review it.

This project aligns perfectly with my expertise in **hardware-aware AI systems, domain-specialized LLM training, and practical deployment optimization**. I am excited to collaborate with the mentors and community to develop an **OpenVINO coder model** that accelerates development workflows and enables efficient local deployment.

Looking forward to the discussion and feedback from everyone here!
## OpenVINO Domain-Specialized Coder Model Architecture

```mermaid
flowchart TB

subgraph DS["1. Data Sources"]
A[OpenVINO Documentation]
B[OpenVINO GitHub Repositories]
C[NNCF Documentation]
D[Intel Developer Guides]
E[Community Sources StackOverflow and GitHub Issues]
end

subgraph DP["2. Data Curation and Dataset Pipeline"]
F[Scraping Engine Python Crawlers]
G[Code Block Extractor]
H[Text Normalization]
I[Duplicate Removal]
J[Quality Filtering]
K[Metadata Annotation Task Type and Hardware]
L[Bug to QA Pair Generator]
M[(OpenVINO Training Dataset 50K Samples)]
end

subgraph TP["3. Model Training Pipeline"]
N[Tokenizer HuggingFace]
O[Base Model Qwen2.5 Coder 7B]

subgraph ME["Memory Efficient Training"]
P[4 bit Quantization BitsAndBytes]
Q[LoRA Adapters PEFT]
R[Gradient Checkpointing]
S[Flash Attention]
end

T[Supervised Fine Tuning]

subgraph GR["GRPO Optimization"]
U[Sample Generator]
V[Reward Function]
W[Code Executability Checker]
X[OpenVINO API Static Analyzer]
Y[Performance Best Practice Detector]
end

Z[Final Fine Tuned Model OpenVINO Coder]

end

subgraph EV["4. Evaluation and Benchmarking"]
AA[OpenVINO Code Bench]
AB[Unit Test Sandbox]
AC[Evaluation Metrics Pass1 Pass5 CodeBLEU]
AD[Human Evaluation]
end

subgraph RAG["5. Retrieval Augmented Knowledge System"]
AE[Documentation Chunking]
AF[Embedding Model BGE]
AG[(Vector Database FAISS or ChromaDB)]
AH[Retriever]
AI[Reranker]
AJ[Context Builder Prompt Templates]
end

subgraph DEP["6. OpenVINO Deployment Pipeline"]
AK[Model Export HF to OpenVINO IR]
AL[Model Optimization FP16 or INT4]
AM[OpenVINO Runtime]
AN[Inference Optimization KV Cache Prefix Cache Batching]
end

subgraph APP["7. Application Layer"]
AO[OpenVINO Code Copilot]
AP[Gradio Chat Interface]
AQ[Developer Prompt OpenVINO Task]
end

A --> F
B --> F
C --> F
D --> F
E --> F

F --> G --> H --> I --> J --> K --> L --> M

M --> N --> O

O --> P --> Q --> R --> S --> T

T --> U --> V

V --> W
V --> X
V --> Y

V --> Z

Z --> AA --> AB --> AC --> AD

A --> AE --> AF --> AG --> AH --> AI --> AJ

AJ --> AO

Z --> AK --> AL --> AM --> AN --> AO

AQ --> AP --> AO

**Comment by MrARwho** [comment]
Hey @Shehrozkashif , 
Your architecture is exceptionally well-designed and showcases a deep understanding of modern, end-to-end LLM pipelines. Choosing Qwen2.5 Coder 7B paired with memory-efficient techniques like 4-bit quantization and Flash Attention ensures a powerful yet hardware-accessible foundation. I am especially impressed by your inclusion of GRPO with domain-specific reward functions—like the OpenVINO API Static Analyzer and Executability Checker—to strongly penalize hallucinations and enforce functional code. Furthermore, "eating your own dog food" by deploying the final model through your own pipeline using OpenVINO IR, INT4, and runtime optimizations like KV caching is a brilliant touch that perfectly validates the ecosystem you are building.

To elevate this pipeline even further, I recommend enhancing your RAG and data curation stages to better handle code semantics and hardware nuances. Standard text chunking often breaks code logic, so implementing Abstract Syntax Tree (AST) parsing for your GitHub data, combined with a Hybrid Search (BM25 + dense embeddings), will drastically improve retrieval accuracy for highly specific API flags. Additionally, injecting synthetic Chain-of-Thought (CoT) reasoning into your training dataset will train the model to explain why it chooses certain optimizations before writing the code. Finally, condition your SFT and GRPO reward functions on specific hardware targets (CPU, GPU, NPU) so the model explicitly learns to suggest device-specific best practices, maximizing OpenVINO's core strengths.

**Comment by Shehrozkashif** [comment]
Thank you @MrARwho, I really appreciate the detailed feedback and the thoughtful suggestions.

Your point about **code-aware chunking** is very insightful. I agree that standard text chunking can easily break code semantics, especially for API-heavy repositories like OpenVINO. Incorporating **AST-based parsing for GitHub code extraction** is a great idea, and it could help preserve structural relationships between functions, classes, and API calls. I will explore integrating AST-based preprocessing for repository data to improve both dataset quality and retrieval performance.

The suggestion to use **hybrid retrieval (BM25 + dense embeddings)** is also very valuable. My initial plan focused on dense embeddings with FAISS/ChromaDB, but combining them with lexical search could significantly improve retrieval of highly specific parameters, API flags, and configuration options that might otherwise be missed by semantic similarity alone.

I also find the idea of incorporating **synthetic Chain-of-Thought (CoT) reasoning** particularly interesting. Including reasoning traces during training could help the model better understand *why* certain OpenVINO optimizations are chosen (for example, performance hints, quantization choices, or device selection) rather than just generating code patterns.

Your suggestion about **hardware-aware conditioning** is also very aligned with OpenVINO’s strengths. Extending the dataset and reward functions to explicitly include **device targets (CPU, GPU, NPU)** and rewarding device-appropriate optimizations could help the model learn to generate more context-aware code.

These are excellent suggestions, and I will definitely explore incorporating them into the proposal and architecture. Thanks again for taking the time to review the design and provide such valuable feedback.

**Comment by Shehrozkashif** [comment]
Hello @7taozhou7 and @yinquan251,

I wanted to kindly follow up regarding my proposal for **GSoC 2026 – Project 6 (OpenVINO Domain-Specialized Coder Model with SFT/GRPO/RAG)**. I shared my proposal via email about a month ago and was hoping to receive your feedback when you have time.

I’m still actively exploring the project ideas and refining the architecture, and I would really appreciate any guidance or suggestions from you to better align my work with the project expectations.

Also, @adrianboguszewski, if possible, could you please help me connect with the mentors or let me know the best way to reach them? I just wanted to make sure my proposal reached them correctly.

Thank you very much for your time and support.
